packages/gatsby/src/db/loki/nodes.js

```diff
@@ -1,12 +1,354 @@
+  invariant(id, `id is null`)
+  invariant(id, `path is null`)
+  const createPageDependency = require(`../../redux/actions/add-page-dependency`)
+  const node = getNode(id)
+  createPageDependency({ path, nodeId: id })
+  return node
+}
+
+/**
+ * Determine if node has changed (by comparing its
+ * `internal.contentDigest`
+ *
+ * @param {string} id
+ * @param {string} digest
+ * @returns {boolean}
+ */
+function hasNodeChanged(id, digest) {
+  const node = getNode(id)
+  if (!node) {
+    return true
+  } else {
+    return node.internal.contentDigest !== digest
+  }
+}
+
+/////////////////////////////////////////////////////////////////////
+// Create/Update/Delete
+/////////////////////////////////////////////////////////////////////
+
+/**
+ * Creates a node in the DB. Will create a collection for the node
+ * type if one hasn't been created yet
+ *
+ * @param {Object} node The node to add. Must have an `id` and
+ * `internal.type`
+ */
+function createNode(node) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let nodeTypeColl = getNodeTypeCollection(type)
+  if (!nodeTypeColl) {
+    nodeTypeColl = createNodeTypeCollection(type)
+  }
+
+  const nodeMetaColl = getDb().getCollection(colls.nodeMeta.name)
+  invariant(nodeMetaColl, `Collection ${colls.nodeMeta.name} should exist`)
+  nodeMetaColl.insert({ id: node.id, typeCollName: nodeTypeColl.name })
+  // TODO what if this insert fails? We will have inserted the id ->
+  // collName mapping, but there won't be any nodes in the type
+  // collection. Need to create a transaction around this
+  return nodeTypeColl.insert(node)
+}
+
+/**
+ * Updates a node in the DB. This works by removing all fields db
+ * fields ($loki, meta, and id) and then using `Object.assign` to copy
+ * all fields in `node` over `oldNode`. Therefore node should contain
+ * ALL fields and not just changes.
+ *
+ * If `oldNode` is not provided, it is assumed that `node` is already
+ * a loki node and we will simply call update on it
+ *
+ * @param {Object} node The new node information. This should be all
+ * the node information. Not just changes
+ * @param {Object} oldNode The old node to write over. Optional.
+ */
+function updateNode(node, oldNode) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = getNodeTypeCollection(type)
+  if (!coll) {
+    invariant(coll, `${type} collection doesn't exist. When trying to update`)
+  }
+
+  if (oldNode) {
+    const lokiKeys = new Set([`$loki`, `meta`, `id`])
+    _.forEach(oldNode, (v, k) => {
+      if (!lokiKeys.has(k)) {
+        delete oldNode[k]
+      }
+    })
+    Object.assign(oldNode, node)
```

Comment:
So, I went down this road because loki doesn't have the concept of `upsert`, i.e if you try and create a node which already has a `$loki` field, or update a field that doesn't, loki will error. Whereas in the redux reducer, we simply overwrite the value for the `node.id` every time. But I just realized I can get around this in `actions.createNode()` by deleting the node itself as well as all its children before we re-insert. See [latest commit](https://github.com/gatsbyjs/gatsby/pull/9919/commits/711af45ede9e997a7b6c7d532c6b404014318401#diff-f61af2b51837961d211c7ea1a4b91837R599). Can you foresee any problems with this approach?